### PR TITLE
Farbling based on browser context

### PR DESCRIPTION
**Pull request by jagadeshjai**
- Migrate |BraveFarblingService| as a keyed service dependent on the profile.
- **Behavioral Change:**  Each Profile has unique farbling's session token and also an incognito session originating from a different profile has its own session token.

# Outcome
[brave_fingerprint_different_profile.webm](https://github.com/user-attachments/assets/0e187544-457a-4749-bd2b-c6adad17c8a9)

Resolves https://github.com/brave/brave-browser/issues/38067
Resolves https://github.com/brave/brave-browser/issues/39182 - partial

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by jagadeshjai** [comment]
@goodov / @ShivanKaul Could you please take a look?

File: browser/net/brave_reduce_language_network_delegate_helper.cc [hunk 1]
```diff
@@ -63,11 +64,15 @@ std::string FarbleAcceptLanguageHeader(
   }
   // Add a fake q value after the language code.
   brave::FarblingPRNG prng;
-  if (g_brave_browser_process->brave_farbling_service()
-          ->MakePseudoRandomGeneratorForURL(
-              origin_url, profile && profile->IsOffTheRecord(), &prng)) {
+
+  auto* brave_farbling_service = BraveFarblingServiceFactory::GetForProfile(
+      profile->IsOffTheRecord() ? profile->GetOriginalProfile() : profile);
```
Comment (goodov) [inline browser/net/brave_reduce_language_network_delegate_helper.cc:69]: you shouldn't pass `profile->GetOriginalProfile()` here. Just use `profile`, you need to configure service factory in a way that the service should understand that it's running under `OffTheRecordProfile`.

to do this pls use `ProfileKeyedServiceFactory` instead of `BrowserContextKeyedServiceFactory` and initialize the factory like so:
```
BraveFarblingServiceFactory::BraveFarblingServiceFactory()
    : ProfileKeyedServiceFactory(
          "BraveFarblingService",
          ProfileSelections::Builder()
              .WithRegular(ProfileSelection::kOwnInstance)
              .WithGuest(ProfileSelection::kOwnInstance)
              .WithSystem(ProfileSelection::kOwnInstance)
              .Build()) {}
```

this way the service should just have `session_token_`.

File: browser/net/brave_reduce_language_network_delegate_helper.cc [hunk 2]
```diff
@@ -63,11 +64,15 @@ std::string FarbleAcceptLanguageHeader(
...
+  if (brave_farbling_service &&
+      brave_farbling_service->MakePseudoRandomGeneratorForURL(
+          origin_url, profile->IsOffTheRecord(), &prng)) {
```
Comment (goodov) [inline browser/net/brave_reduce_language_network_delegate_helper.cc:72]: here you shouldn't pass `profile->IsOffTheRecord()` for the same reason.

File: browser/net/brave_reduce_language_network_delegate_helper.cc (hunk 1 above)
Comment (jagadeshjai) [reply browser/net/brave_reduce_language_network_delegate_helper.cc:69]: ohh, I didn't know about this, Thanks!

Comment (jagadeshjai) [reply browser/net/brave_reduce_language_network_delegate_helper.cc:69]: BTW, Do we really need to create the service for System profile?
@goodov

Comment (jagadeshjai) [reply browser/net/brave_reduce_language_network_delegate_helper.cc]: For now, I'm not creating service for system profile.

Comment (goodov) [reply browser/net/brave_reduce_language_network_delegate_helper.cc]: > For now, I'm not creating service for system profile.

as long as the `GetForProfile` result is checked before using, this is fine. I think the system profile is used on profile creation window and some other parts like this (when there's no profile yet).

Comment (jagadeshjai) [reply browser/net/brave_reduce_language_network_delegate_helper.cc]: Yes, as per my knowledge it will used in profile picker as well.

@goodov Could you please help with CI for this PR?

**Comment by goodov** [comment]
The change looks good, but is missing additional tests to ensure the session token is different in normal/incognito profile, and is different in another profile.

Ideally tests should run a simple fingerprinting check (which we farble) that should return a different value for each profile.

**Comment by jagadeshjai** [comment]
> The change looks good, but is missing additional tests to ensure the session token is different in normal/incognito profile, and is different in another profile.
> 
> Ideally tests should run a simple fingerprinting check (which we farble) that should return a different value for each profile.

Sure! Will add that :+1:

**Comment by goodov** [comment]
please also rebase to the latest `master`.

image

**Comment by jagadeshjai** [comment]
@goodov 

Since we are checking whether current process has `kTestType` in https://github.com/brave/brave-core/blob/6399383194d6c6309c18e6370899239c0b48f1d1/browser/brave_content_browser_client.cc#L893 before setting up the session token for current profile, We can't write a test by checking the farblied values, instead we can compare two different profiles session tokens. 

Will that be enough for the testing?

**Comment by goodov** [comment]
> @goodov
> 
> Since we are checking whether current process has `kTestType` in
> 
> https://github.com/brave/brave-core/blob/6399383194d6c6309c18e6370899239c0b48f1d1/browser/brave_content_browser_client.cc#L893
> 
> before setting up the session token for current profile, We can't write a test by checking the farblied values, instead we can compare two different profiles session tokens.
> Will that be enough for the testing?

can you try to go this way? https://gist.github.com/goodov/57e7b932d48093a417b4b5aa7a3a1058

**Comment by Ghxst** [comment]
Apologies if my comment is out of place here, just wanted to ask if this works for all contexts (related to https://github.com/brave/brave-browser/issues/39182) where we generate contexts in headless environments through [createBrowserContext](https://chromedevtools.github.io/devtools-protocol/tot/Target/#method-createBrowserContext). If you need help with steps to test this or anything else please let me know.

**Comment by jagadeshjai** [comment]
> can you try to go this way? https://gist.github.com/goodov/57e7b932d48093a417b4b5aa7a3a1058


Okay, but if we do it like this, then we might miss the intention of this https://github.com/brave/brave-core/blob/6399383194d6c6309c18e6370899239c0b48f1d1/third_party/blink/renderer/core/farbling/brave_session_cache.cc#L218
right ? @goodov

**Comment by jagadeshjai** [comment]
> Apologies if my comment is out of place here, just wanted to ask if this works for all contexts (related to [brave/brave-browser#39182](https://github.com/brave/brave-browser/issues/39182)) where we generate contexts in headless environments through [createBrowserContext](https://chromedevtools.github.io/devtools-protocol/tot/Target/#method-createBrowserContext). If you need help with steps to test this or anything else please let me know.

Yes, AFICT, it will be unique for each headless (browser context).

**Comment by goodov** [comment]
> > can you try to go this way? https://gist.github.com/goodov/57e7b932d48093a417b4b5aa7a3a1058
> 
> Okay, but if we do it like this, then we might miss the intention of this
> 
> https://github.com/brave/brave-core/blob/6399383194d6c6309c18e6370899239c0b48f1d1/third_party/blink/renderer/core/farbling/brave_session_cache.cc#L218
> 
> 
> right ? @goodov

not sure how this may be related. The code you linked ensures the tests will fail if we break the `kBraveSessionToken` flag passing accidentally.

The flag is always passed, even if `kTestType` is set (before my proposed change and after). It's just the value is set to `12345` if the test is currently active. My change makes it possible to increase this value consistently for each new service instance.

**Comment by jagadeshjai** [comment]
> The flag is always passed, even if `kTestType` is set (before my proposed change and after). It's just the value is set to `12345` if the test is currently active. My change makes it possible to increase this value consistently for each new service instance.


Okay, got it! But I still have a doubt: how will faking the session token during testing (i.e., incrementing it) ensure the real-time scenario of having different session tokens based on the browser context?

**Comment by goodov** [comment]
> > The flag is always passed, even if `kTestType` is set (before my proposed change and after). It's just the value is set to `12345` if the test is currently active. My change makes it possible to increase this value consistently for each new service instance.
> 
> Okay, got it! But I still have a doubt: how will faking the session token during testing (i.e., incrementing it) ensure the real-time scenario of having different session tokens based on the browser context?

Actually, I'm okay with keeping the tests you added, it should be fine for now. I work on a full replacement of this thing, so the tests anyways will be rewritten.

Can you please rebase the change to latest master so I can run CI?

**Comment by jagadeshjai** [comment]
> Can you please rebase the change to latest master so I can run CI?

Now I'll do it, But why can't we automate the rebase while running the CI via Brave-builds?

If we do so, it will prevent contributors from building with additional changes. (I believe Chromium has this), Because while rebasing, most of the time in my experience, its building more than 25k targets takes around 5-6 hours to complete. 

However, if we need to resolve merge conflicts or if the CI fails, contributors can manually rebase. @goodov 


CC: DevOPs @mihaiplesa

File: third_party/blink/renderer/core/farbling/brave_session_cache.cc [hunk 3]
```diff
@@ -379,8 +393,9 @@ bool BraveSessionCache::AllowFontFamily(
     case BraveFarblingLevel::BALANCED:
     case BraveFarblingLevel::MAXIMUM: {
       if (AllowFontByFamilyName(family_name,
-                                blink::DefaultLanguage().GetString().Left(2)))
+                                blink::DefaultLanguage().GetString().Left(2))) {
         return true;
+      }
```
Comment (goodov) [inline third_party/blink/renderer/core/farbling/brave_session_cache.cc:398]: please revert unnecessary changes. This file should stay as is.

File: components/brave_shields/content/test/brave_farbling_service_unittest.cc [hunk 4]
```diff
@@ -3,19 +3,20 @@
...
 // You can obtain one at https://mozilla.org/MPL/2.0/.
 
+#include "brave/components/brave_shields/content/browser/brave_farbling_service.h"
+
 #include <memory>
 #include <tuple>
 
-#include "brave/components/brave_shields/content/browser/brave_farbling_service.h"
+#include "chrome/test/base/testing_profile.h"
+#include "content/public/test/browser_task_environment.h"
```
Comment (goodov) [inline components/brave_shields/content/test/brave_farbling_service_unittest.cc:12]: please cleanup. You don't need `testing_profile`.

Not sure why you need `BrowserTaskEnvironment`.

File: third_party/blink/renderer/core/farbling/brave_session_cache.cc (hunk 3 above)
Comment (jagadeshjai) [reply third_party/blink/renderer/core/farbling/brave_session_cache.cc]: Reveted.

File: components/brave_shields/content/test/brave_farbling_service_unittest.cc (hunk 4 above)
Comment (jagadeshjai) [reply components/brave_shields/content/test/brave_farbling_service_unittest.cc]: Removed. It was used when I passed the profile to `BraveFarblingService` before.

**Comment by goodov** [comment]
```
+ npm run presubmit
Running Python 3 presubmit upload checks ...
/home/ubuntu/workspace/teamcity/work/contributor-jagadeshjai-feature_farbling_session_key_based_on_profile/src/brave/browser/farbling/brave_farbling_based_on_profiles_browsertest.cc:0:  (cpplint) No copyright message found.  You should have a line: "Copyright [year] <Copyright Owner>"  [legal/copyright] [5]
  checking for commit objects in tree took a long time: 0.5s
** Presubmit ERRORS: 2 **
Changelist failed cpplint.py check. Search the output for "(cpplint)"
Expected one of license headers in new files:
// Copyright (c) 2024 The Brave Authors. All rights reserved.
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this file,
// You can obtain one at https://mozilla.org/MPL/2.0/
/* Copyright (c) 2024 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */
# Copyright (c) 2024 The Brave Authors. All rights reserved.
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at https://mozilla.org/MPL/2.0/
  browser/farbling/brave_farbling_based_on_profiles_browsertest.cc
Presubmit checks took 2.6s to calculate.
There were Python 3 presubmit errors.
null
Process exited with code 1
```

**Comment by jagadeshjai** [comment]
> ```
> + npm run presubmit

Done. @goodov

**Review by goodov — APPROVED** [review]
lgtm. can be merged cc @bsclifton

**Comment by bsclifton** [comment]
Kicked off CI to make sure - https://github.com/brave/brave-core/pull/24669. Will merge if that's good to go 😄

File: third_party/blink/renderer/core/farbling/brave_session_cache.cc [hunk 5]
```diff
@@ -214,9 +214,12 @@ BraveSessionCache::BraveSessionCache(ExecutionContext& context)
   CHECK(h.Init(reinterpret_cast<const unsigned char*>(&session_key_),
                sizeof session_key_));
   CHECK(h.Sign(domain, domain_key_, sizeof domain_key_));
-  settings_client_ = GetContentSettingsClientFor(&context, true);
-  if (settings_client_ != nullptr) {
-    auto raw_farbling_level = settings_client_->GetBraveFarblingLevel(
+  const uint64_t* fudge = reinterpret_cast<const uint64_t*>(domain_key_);
```
Comment (jagadeshjai) [inline third_party/blink/renderer/core/farbling/brave_session_cache.cc:217]: Changes in this file not meant to be here, need to revert it.
CC: @bsclifton

Comment (jagadeshjai) [reply third_party/blink/renderer/core/farbling/brave_session_cache.cc]: Reverted it as clean up commit(Since same file mentioned in two commit, so ignored `git revert`, now you can start the CI.

@bsclifton

**Comment by bsclifton** [comment]
Thanks, @jagadeshjai! Got a new run taking place now 😄

**Comment by bsclifton** [comment]
CI looks good! 😄
